haskell/lint.bzl

```diff
@@ -0,0 +1,170 @@
+"""Lint Haskell sources using hlint."""
+
+HaskellHLintInfo = provider(
+    doc = "Provider that collects files produced by hlint",
+    fields = {
+        "outputs": "depset of hlint log files.",
+    },
+)
+
+def _collect_hlint_logs(deps):
+    return depset(transitive = [
+        dep[HaskellHLintInfo].outputs
+        for dep in deps
+        if HaskellHLintInfo in dep
+    ])
+
+def _haskell_lint_aspect_impl(target, ctx):
+    hlint_toolchain = ctx.toolchains["@io_tweag_rules_haskell//haskell/lint:toolchain"]
+    hint = hlint_toolchain.hint
+
+    inputFiles = []
+    inputPaths = []
+    if hasattr(ctx.rule.attr, "srcs"):
+        for src in ctx.rule.attr.srcs:
+            for f in src.files:
+                # We want to only do native Haskell source files, which
+                # seems to involve ignoring these generated paths
+                # (the f.is_source almost always returns True)
+                if all([
+                    f.path.endswith(".hs"),
+                    f.path.startswith("external/") == False,
+                    f.path.startswith("bazel-out/") == False,
+                    f.path.startswith("nix/") == False,
+                ]):
+                    inputFiles.append(f)
+                    inputPaths.append(f.path)
```

Comment:
What happens to generated Haskell files? Normally I’d expect they should not be linted by default.

Shouldn’t we restrict to a specified subset of rules instead of trying to heuristically filter sources?

Reply:
Generated files are filtered out. Now using `f.is_source`. I've now also restricted the aspect to targets providing `HaskellInfo`.